Control to Manage temporary, time-bound IAM roles

ashishjain2412 opened the pull request:
The  ChaosGCP IAM Policy Control  allows you to manage temporary, time-bound IAM roles for a specified project and members. This control grants specified IAM roles to the provided members, with an expiration time determined by the given parameter. It can be used to grant and revoke IAM roles to/from specified members before and after an experiment, respectively.
This control can be used to improve the security of Chaos Toolkit experiments by ensuring that IAM roles are only granted for the duration of the experiment.
The ChaosGCP IAM Policy Control is Python-based and designed to be used in conjunction with Chaos Toolkit experiments to grant and revoke IAM roles .

ashishjain2412 committed:
Added chaosgcp.iam.controls.policy control to Manage temporary, time-bound IAM roles for the specified project and members

Signed-off-by: Ashish Jain <ashishjn@google.com>

ashishjain2412 committed:
Added Test Cases for chaosgcp.iam.controls.policy control

Signed-off-by: Ashish Jain <ashishjn@google.com>

ashishjain2412 committed:
Added Readme for Google cloud 's IAM policy control

Signed-off-by: Ashish Jain <ashishjn@google.com>

ashishjain2412 committed:
modified changelog

Signed-off-by: Ashish Jain <ashishjn@google.com>

ashishjain2412 committed:
Reformatted after PDM format

Signed-off-by: Ashish Jain <ashishjn@google.com>

ashishjain2412 committed:
added iam init.py

Signed-off-by: Ashish Jain <ashishjn@google.com>

ashishjain2412 committed:
Modified iam control readme

Signed-off-by: Ashish Jain <ashishjn@google.com>

ashishjain2412 committed:
Modified changelog

Signed-off-by: Ashish Jain <ashishjn@google.com>

Lawouach commented on CHANGELOG.md:
```diff
@@ -1,5 +1,28 @@
 # Changelog
 
+#### Added  2024-04-26
```
No date needed on these sections. Only used per release section header that the PR doesn't set.

Lawouach replied:
The section should be insert after the `## [Unreleased][]` section not before.

ashishjain2412 replied:
removed date , section added after unreleased.

Lawouach commented on CHANGELOG.md:
```diff
@@ -1,5 +1,28 @@
+#### Added  2024-04-26
+
+* The `chaosgcp.iam.controls.policy` control to Manage temporary, time-bound IAM roles for the specified project and members.
+* This controls grants specified IAM roles to the provided members, with an expiration time determined by the given expiry_time_in_minutes  before experiment and revokes the specified IAM roles from the provided members after the experiment:
```
Try to keep each column under a 80-width rule. I know we have bigger screens but I think it's still easier to read with narrower width.

ashishjain2412 replied:
modified changelog with 80-width rule

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+# Copyright 2024 Google LLC
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
While I'm happy for the copyright line to be set, I'm not sure we need to repeat the LICENSE here since it's covered by the one at the top of the repository.

ashishjain2412 replied:
Based on Guidelines, It should have complete copyright notice without any modifications .Kindly let me know if that is ok .

Lawouach replied:
Okay, that's fine.

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+import datetime
+import time
+import logging
+from datetime import timezone
```
These need to be sorted as follows:


```
# Python stdlib imports

# third party imports

# local package imports
```

Here this would look like:

```
import datetime
import logging
import time
from datetime import datetime
from typing import List

import google.auth
import googleapiclient.discovery
from chaoslib.exceptions import InterruptExecution
from chaoslib.types import Configuration, Experiment, Journal, Secrets
```


In theory, running `pdm run format` might help.

ashishjain2412 replied:
Thanks , I have rearranged them .

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+import logging
+from datetime import timezone
+
+logger = logging.getLogger("chaostoolkit")
```
this should go after the `__all__ = ...` statement

ashishjain2412 replied:
Moved after __all__ statement

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+
+__all__ = ["before_experiment_control", "after_experiment_control"]
+
+set_current_time = datetime.datetime.now(timezone.utc)
```
this will be instanciated once only during the first import of this module. is that the expected outcome? if so, please add a comment explaining this variable :)

ashishjain2412 replied:
Yes that is expected, it should be constant . I have added the comment .

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+def before_experiment_control(
+    context: Experiment,
+    project_id: str,
+    roles: List,
```
can you be more explicit on the typing here, maybe `List[str]`?

ashishjain2412 replied:
added list[str]

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+    context: Experiment,
+    project_id: str,
+    roles: List,
+    members: List,
```
can you be more explicit on the typing here, maybe `List[str]`?

ashishjain2412 replied:
added list[str]

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+    members: List,
+    configuration: Configuration = None,
+    secrets: Secrets = None,
+    iam_propogation_sleep_time_in_minutes=2,
+    expiry_time_in_minutes=10,
```
by convention, we always end the signature withe the configuration and secrets. Python and chaostoolkit won't care, but this make code readability more consistent.

Lawouach replied:
Also make sure to add a type for both arguments

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+    expiry_time_in_minutes=10,
+) -> None:
+    """
+    to ADD IAM roles
```
Can you expand the docstring a bit more? This is useful in when we generate the main documentation.

ashishjain2412 replied:
Added extended docstring

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+    project_id: str,
+    roles: List,
+    members: List,
+    iam_propogation_sleep_time_in_minutes=2,
+    expiry_time_in_minutes=10,
```
add typing annotation

ashishjain2412 replied:
Added Type

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+    context: Experiment,
+    state: Journal,
+    project_id: str,
+    roles: List,
+    members: List,
```
be more precise on typing annotation

ashishjain2412 replied:
Added List[str]

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+    secrets: Secrets = None,
+) -> None:
+    """
+    to remove IAM roles
```
extend docstring

ashishjain2412 replied:
Added extended docstring

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+
+def manage_time_bound_temporary_iam_roles(
+    project_id: str,
+    members: list[str],
+    roles: list[str],
+    roles_iam_condition: dict[str, str],
```
Please switch to `Dict` instead of `dict`. The latter really only came into play with Python 3.9 but we support 3.8.

https://docs.python.org/3.12/library/typing.html#typing.Dict

ashishjain2412 replied:
changed to Dict

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+    roles_iam_condition: dict[str, str],
+    manage_type: str,
+    iam_propogation_sleep_time_in_minutes: int = 2,
+) -> None:
```
is there nothing useful to return from this function?

ashishjain2412 replied:
No

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+            manage_type="add",
+        )
+    """
+    # Type checks for arguments
+    if not isinstance(project_id, str):
+        raise TypeError("project_id must be a string")
+    if not isinstance(members, list):
+        raise TypeError("members must be a list of strings")
+    if not isinstance(roles, list):
+        raise TypeError("roles must be a list of strings")
+    if not isinstance(roles_iam_condition, dict):
+        raise TypeError("roles_iam_condition must be a dictionary")
+    if not isinstance(manage_type, str):
+        raise TypeError("manage_type must be a string")
+    if not isinstance(iam_propogation_sleep_time_in_minutes, int):
```
in theory this isn't really needed as `chaos validate` should care for this check. worth verifying if that's the case.

ashishjain2412 replied:
removed

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+        raise ValueError("Invalid manage type: must be 'add' or 'remove'")
+
+
+def initialize_service() -> dict:
```
same comment on using `Dict` vs `dict`

ashishjain2412 replied:
changed to Dict

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+def initialize_service() -> dict:
+    """Initializes a Cloud Resource Manager service."""
+
+    credentials, _ = google.auth.default(
+        scopes=["https://www.googleapis.com/auth/cloud-platform"]
+    )
+    crm_service = googleapiclient.discovery.build(
+        "cloudresourcemanager", "v1", credentials=credentials
+    )
```
Interesting. Why not use the `chaosgcp.__init__.client` function here?

ashishjain2412 replied:
Modified the policy.py with chaosgcp.__init__.client

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+
+def get_policy(crm_service, project_id: str, version: int = 3) -> dict:
+    """Gets IAM policy for a project."""
+    logger.info("Getting IAM policy for a project.")
```
We tend to keep the `info` level only for experiment related necessary logs. I don't mind that we use `info` but if this log is more "debug" oriented, then it should be `.debug(...)` instead.

ashishjain2412 replied:
It is info related , because user should know about the policy changes and can act accordingly if something goes wrong .

Lawouach commented on chaosgcp/iam/controls/policy.py:
```diff
@@ -0,0 +1,304 @@
+def modify_policy_add_roles(
+    crm_service,
+    project_id: str,
+    roles: list[str],
+    member: list[str],
+    iam_condition: dict[str, str],
+    iam_propogation_sleep_time_in_minutes: int,
```
fix the typing annotation to be supported by Python 3.8+

ashishjain2412 replied:
changed to List and Dict .

Lawouach commented on chaosgcp/iam/controls/readme.md:
```diff
@@ -0,0 +1,27 @@
+## The `chaosgcp.iam.controls.policy` control to Manage temporary, time-bound IAM roles for the specified project and members.
```
We don't usually put readme in subpackage. I'd rather we keep at the top-level readme level.

ashishjain2412 replied:
This was dedicated to iam control only . I have renamed to policy.md to keep it specific to policy control. I found in top level readme.MD nothing specific was mentioned related to controls. Kindly Let me know if you want to move content of policy.md to readme.MD (top level ) 
https://github.com/ashishjain2412/chaostoolkit-google-cloud-platform/blob/2a4dc9f5026ce54e88eb0925291c83e921b4662b/chaosgcp/iam/controls/policy.md

Lawouach replied:
I think that's fine then :)

Lawouach commented on tests/test_iam_controls_policy.py:
```diff
@@ -0,0 +1,94 @@
+# Copyright 2024 Google LLC
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
remove in favour of the LICENSE file  governing this repository

ashishjain2412 replied:
Based on Internal Guidelines, It should have complete copyright notice without any modifications .Kindly let me know if that is ok .

Lawouach reviewed (changes requested):
Thank yo ufor this PR Ashish!

I've made a few inline comments. Also, I had a couple of questions:

* Is there any concurrency risks between multiple simultaneous calls to the same experiment?
* Are the policies change shown in any audit traces on GCP? If so, could we document this so operators know where to look for them if needed
* What are the necessary permissions for the user calling this experiment so that the grants are allowed?

ashishjain2412 committed:
Merge branch 'master' into master

ashishjain2412 committed:
Changed the policy.py as per the PR comments

Signed-off-by: Ashish Jain <ashishjn@google.com>

added docstring

Signed-off-by: Ashish Jain <ashishjn@google.com>

added docstring

Signed-off-by: Ashish Jain <ashishjn@google.com>

removed type checking and add type

Signed-off-by: Ashish Jain <ashishjn@google.com>

Added comments on set_current_time

Signed-off-by: Ashish Jain <ashishjn@google.com>

Modified policy.md added info of handling IAM propogation delay

Signed-off-by: Ashish Jain <ashishjn@google.com>

Modified control with default client, added audit logs related information

Signed-off-by: Ashish Jain <ashishjn@google.com>

Modified policy.md

Signed-off-by: Ashish Jain <ashishjn@google.com>

Added Prerequisites section in readme , chnaged the name to policy.md

Signed-off-by: Ashish Jain <ashishjn@google.com>

changed list and dict to python3.8+ supported

Signed-off-by: Ashish Jain <ashishjn@google.com>

sorted the import statements and logger

Signed-off-by: Ashish Jain <ashishjn@google.com>

modified changelog with 80-width rule and removed date , section added after unreleased

Signed-off-by: Ashish Jain <ashishjn@google.com>

ashishjain2412 commented:
> Thank yo ufor this PR Ashish!
> 
> I've made a few inline comments. Also, I had a couple of questions:
> 
> * Is there any concurrency risks between multiple simultaneous calls to the same experiment?
> * Are the policies change shown in any audit traces on GCP? If so, could we document this so operators know where to look for them if needed
> * What are the necessary permissions for the user calling this experiment so that the grants are allowed?

Added Audit related info and required permission in policy.md. And no concurrency risk because each experiment will have different set of conditional roles .

Lawouach commented:
Hey @ashishjain2412 thanks for the changes. Could you squash your changes and push? I'm finding it difficult to make my review.

ashishjain2412 commented:
> Hey @ashishjain2412 thanks for the changes. Could you squash your changes and push? I'm finding it difficult to make my review.



> Hey @ashishjain2412 thanks for the changes. Could you squash your changes and push? I'm finding it difficult to make my review.

@Lawouach Thank you , I have squash my recent changes to one. Kindly review .

Lawouach commented:
Thanks. though I'm confused. It says your own branch has more commits. I can't really review any of your recent changes in this repo properly, they don't show up in https://github.com/chaostoolkit-incubator/chaostoolkit-google-cloud-platform/pull/23/files

That has made the review quite difficult.

ashishjain2412 commented:
> Thanks. though I'm confused. It says your own branch has more commits. I can't really review any of your recent changes in this repo properly, they don't show up in https://github.com/chaostoolkit-incubator/chaostoolkit-google-cloud-platform/pull/23/files
> 
> That has made the review quite difficult.

Hi @Lawouach , could you please retry , i changed the base again, if it sync issue.
kindly let me know .

Lawouach commented:
Thanks @ashishjain2412 but I'd rather you let me know if the current view of the PR represents your entire changeset? I feel it's complete but I'd rather you confirm it.

ashishjain2412 commented:
Yes it is complete. Thanks!

Lawouach commented:
great. I'll merge it then :)
